### DISCUSSION TITLE
how to implement tab drawing in python?

### DISCUSSION THREAD

**Discussion by zzhaolei** [kovidgoyal/kitty#4366] [Q&A] [answered]
I can't wait to experience new features. I’ve switched to the nightly version.

I see support for drawing tabs in Python, But I don’t know where to write code (I use Python in my work) .

Can you give me some pointers？

**Comment by page-down** [comment, marked as answer]
https://github.com/kovidgoyal/kitty/blob/master/kitty/tab_bar.py

See `draw_tab_with_*`, e.g. `draw_tab_with_separator()`.

~/.config/kitty/tab_bar.py

```python
from kitty.fast_data_types import Screen
# from kitty.rgb import Color
# from kitty.utils import color_as_int

from kitty.tab_bar import DrawData, ExtraData, TabBarData
# from kitty.tab_bar import as_rgb, draw_title


def draw_tab(
    draw_data: DrawData, screen: Screen, tab: TabBarData,
    before: int, max_title_length: int, index: int, is_last: bool,
    extra_data: ExtraData
) -> int:
    # Draw your own tab with `screen.draw()`, `screen.apply_sgr()`.
    # You can use the function `draw_title(draw_data, screen, tab, index)`.
    # Move the cursor by assigning the value to `screen.cursor.x`.
    # Set the color and style with `screen.cursor.fg: int, bg: int, bold: bool, italic: bool`.
    # Returns the position of the cursor after the tab has been drawn.
    end = screen.cursor.x
    return end
```

**Reply by zzhaolei** [reply]
It’s not working for me.

That’s what I wrote:
`~/.config/kitty/tab_bar.py`
```python
from kitty.fast_data_types import Screen
from kitty.rgb import Color
from kitty.tab_bar import DrawData, ExtraData, TabBarData, as_rgb, draw_title
from kitty.utils import color_as_int


def draw_tab_with_separator(
    draw_data: DrawData, screen: Screen, tab: TabBarData,
    before: int, max_title_length: int, index: int, is_last: bool,
    extra_data: ExtraData
) -> int:
    screen.draw("test")
    draw_title(draw_data, screen, tab, index)

    end = screen.cursor.x
    return end
```

**Reply by page-down** [reply]
Function with the name `draw_tab` are required. Don't copy the name `draw_tab_with_separator`.

What I mean is that you can refer to existing drawing functions, but don't copy the names as well.

**Reply by zzhaolei** [reply]
I don't understand. Can you write an example?

**Reply by page-down** [reply]
> That’s what I wrote:
> ...
> def draw_tab_with_separator(

Change to `def draw_tab(`

**Reply by zzhaolei** [reply]
I tried, but it didn’t work.

```Python
from kitty.fast_data_types import Screen
from kitty.tab_bar import DrawData, ExtraData, TabBarData, as_rgb, draw_title

# from kitty.rgb import Color
# from kitty.utils import color_as_int

def draw_tab(
    draw_data: DrawData, screen: Screen, tab: TabBarData,
    before: int, max_title_length: int, index: int, is_last: bool,
    extra_data: ExtraData
) -> int:
    screen.draw("test")
    tab.title = "test1"
    draw_title(draw_data, screen, tab, 5)
    end = screen.cursor.x
    return end
```

**Reply by page-down** [reply]
Run a separate kitty process and fix the code according to the stdout ,stderr error message.

```shell
kitty --config=NONE -o tab_bar_style=custom
```

You can't set `tab.title = 'title'`.

The examples are already in `kitty/tab_bar.py`. The link is in the first line of my answer. All the `draw_tab_with_*()` functions are `examples`.

**Reply by zzhaolei** [reply]
I'm so, so sorry. I didn’t set it `tab_bar_style=custom`, Now it works.

My sincere apologies.

**Comment by horseinthesky** [comment]
Hello.
https://github.com/kovidgoyal/kitty/blob/master/kitty/tab_bar.py is quite huge.

Can't find this feature's documentation on official page. Could you provide minimal config to make tabs edges rounded? So basically speaking I want to be able to change tabs surrounding symbols.

**Reply by page-down** [reply]
> Can't find this feature's documentation on official page.

https://sw.kovidgoyal.net/kitty/conf/#opt-kitty.tab_bar_style

However only the config options are described. As for how to draw, you need to check the drawing functions in that file, there is no detailed documentation with tutorials for each function yet.

> ... I want to be able to change tabs surrounding symbols.

https://github.com/kovidgoyal/kitty/discussions/3984#discussioncomment-1300900

Is this what you want?

**Reply by horseinthesky** [reply]
> However only the config options are described. As for how to draw, you need to check the drawing functions in that file, there is no detailed documentation with tutorials for each function yet.
> 
> >

I've found this discussion that has pretty much the exact way I want to see my tabs.
https://github.com/kovidgoyal/kitty/discussions/4447#discussioncomment-1922580

As far as I understand this feature (to do it with python) is very new so would be nice to see some docs about classes and functions used in the future. For now, I can play with it and see how it affects the end result. Great to have control of this with Python. Thank you.
